### PR TITLE
Enable `text-offset` with variable label placement

### LINKED ISSUE #8598: text-radial-offset does not account for icon size or bounding boxes
Stub ticket to capture customer feedback:

`text-radial-offset` does not take into account symbol icon dimensions. Text with `text-radial-offset` may overlap wide or horizontally stretched icons with a `"*-left"`/`"*-right"` `text-variable-anchor` position.

text_right_shift_broken

cc @pozdnyakov @alexshalamov

### PR DESCRIPTION
**Pull request by pozdnyakov**
This PR allows using of `text-offset` together with `text-variable-anchor`, if `text-radial-offset` is not provided. Assuming `text-offset` values are [x, y] the offset is calculated based on the `text-variable-anchor` values as following:

"left": [-x, 0]
"right": [x, 0]
"top": [0, -y]
"bottom": [0, y]
"top-left": [-x, -y]
"top-right": [x, -y]
"bottom-left": [-x, y]
"bottom-right": [x, y]

download

We allow only positive `text-offset` values for simplicity.

This is an alternate approach to https://github.com/mapbox/mapbox-gl-js/pull/8609

Fixes #8598

### CODE REVIEW

**Comment by asheemmamoowala** [comment]
cc @mapbox/map-design-team @mapbox/studio

**Comment by pozdnyakov** [comment]
> In addition to the docs change, this is also likely to break styles where both text-variable-anchor has been specified without text-radial-offset. 

The advantage of this approach is that it won't break the existing styles, as at the moment using of `text-offset` together with `text-variable-anchor` [is not allowed by the style specification](https://docs.mapbox.com/mapbox-gl-js/style-spec/#layout-symbol-text-offset) :smile:

**Comment by pozdnyakov** [comment]
> This will need an update to the style-spec and their docs as well before merge.

Done in the latest commit, please take a look

File: src/symbol/symbol_layout.js [hunk 1]
```diff
@@ -510,7 +550,11 @@ function addSymbol(bucket: SymbolBucket,
     let numVerticalGlyphVertices = 0;
     const placedTextSymbolIndices = {};
     let key = murmur3('');
-    const radialTextOffset = (layer.layout.get('text-radial-offset').evaluate(feature, {}) || 0) * ONE_EM;
+    let radialTextOffset0 = (layer.layout.get('text-radial-offset').evaluate(feature, {}) || 0) * ONE_EM;
+    let radialTextOffset1 = Number.POSITIVE_INFINITY;
+    if (!radialTextOffset0) {
```
Comment (alexshalamov) [inline src/symbol/symbol_layout.js]: Would it be better to check for undefined? Right now, undefined and `text-radial-offset: 0` behave in the same way, falling back to `text-offset`.

**Comment by ansis** [comment]
I'd be in favour of remove the two-component radial offset if we implement this for text-offset. I'll take another closer look at the implementation after a studio/map design review but it looks good!

**Comment by dereklieu** [comment]
> at the moment using of text-offset together with text-variable-anchor is not allowed by the style specification

I don't believe it's not allowed, instead having one means the other is ignored. In Studio, if you set text-offset, then set text-variable-anchor, the text-offset value is not removed from your stylesheet. 

To @asheemmamoowala's point, this does have the potential to break some maps and we would need to call it out. The impact may be somewhat contained as radial offset is a relatively new feature in Studio.

The changes to the style spec look good to me.

**Comment by asheemmamoowala** [comment]
> @ansis:  I'll take another closer look at the implementation after a studio/map design review but it looks good!

> @dereklieu : The changes to the style spec look good to me.

@ansis could you take a look at the implementation?

File: src/symbol/symbol_layout.js (hunk 1 above)
Comment (pozdnyakov) [reply src/symbol/symbol_layout.js]: Fixed in the latest commit, thanks!

File: test/integration/render-tests/text-variable-anchor/all-anchors-offset-zero/style.json [hunk 2]
```diff
@@ -0,0 +1,122 @@
+{
+  "version": 8,
+  "metadata": {
+    "test": {
+      "height": 256,
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply test/integration/render-tests/text-variable-anchor/all-anchors-offset-zero/style.json]: Thanks for noticing! 64x64 is too small, but 128x128 fits very well

File: src/symbol/symbol_layout.js [hunk 3]
```diff
@@ -64,46 +64,86 @@ export type TextAnchor = 'center' | 'left' | 'right' | 'top' | 'bottom' | 'top-l
...
 
-export function evaluateRadialOffset(anchor: TextAnchor, radialOffset: number) {
-    let x = 0, y = 0;
-    // solve for r where r^2 + r^2 = radialOffset^2
-    const hypotenuse = radialOffset / Math.sqrt(2);
+export function evaluateRadialOffset(anchor: TextAnchor, offset: [number, number]) {
+
+    function fromRadialOffset(anchor: TextAnchor, radialOffset: number) {
+        let x = 0, y = 0;
+        if (radialOffset < 0) radialOffset = 0; // Ignore negative offset.
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply src/symbol/symbol_layout.js]: Indeed, it's better to use absolute value :+1:

**Comment by pozdnyakov** [comment]
@asheemmamoowala @chloekraw thanks for your comments! Applied in the latest commits.

File: src/style-spec/reference/v8.json [hunk 4]
```diff
@@ -2214,7 +2214,7 @@
     },
     "text-offset": {
       "type": "array",
-      "doc": "Offset distance of text from its anchor. Positive values indicate right and down, while negative values indicate left and up.",
+      "doc": "Offset distance of text from its anchor. Positive values indicate right and down, while negative values indicate left and up. Note that in combination with `text-variable-anchor` absolute values are used and they indicate radial offsets along x- and y-axis accordingly.",
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply src/style-spec/reference/v8.json]: Thanks for your proposal!

File: src/data/bucket/symbol_attributes.js [hunk 5]
```diff
@@ -96,7 +96,7 @@ export const symbolInstance = createLayout([
     { type: 'Uint16', name: 'numIconVertices' },
     { type: 'Uint32', name: 'crossTileID' },
     { type: 'Float32', name: 'textBoxScale'},
-    { type: 'Float32', name: 'radialTextOffset'}
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply src/data/bucket/symbol_attributes.js]: Maybe `textOffset` is even better, once assigned, it is not variable any more here :thinking:

File: src/symbol/symbol_layout.js [hunk 6]
```diff
@@ -64,46 +64,86 @@ export type TextAnchor = 'center' | 'left' | 'right' | 'top' | 'bottom' | 'top-l
...
+        case 'bottom-left':
+        case 'left':
+            x = offsetX;
+            break;
+        }
+
+        return [x, y];
     }
 
-    return [x, y];
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply src/symbol/symbol_layout.js]: I've introduced `const INVALID_TEXT_OFFSET = Number.POSITIVE_INFINITY;`, hope it will improve readability

**Pull request merged**